src/routes/v2/sgidAuth.ts

```diff
@@ -86,25 +90,130 @@ export class SgidAuthRouter {
 
     const { nonce, codeVerifier } = cookieData
 
-    // Exchange the authorization code and code verifier for the access token, then use the access token to retrieve user's email
-    await this.sgidAuthService
+    // Exchange the authorization code and code verifier for the access token, then use the access token to retrieve user's data
+    const userDataRes = await this.sgidAuthService
       .retrieveSgidAccessToken({ authCode, nonce, codeVerifier })
       .andThen(({ accessToken, sub }) =>
         // We can immediately process the access token - we only need to retrieve the email from sgid
-        this.sgidAuthService.retrieveSgidUserEmail(accessToken, sub)
+        this.sgidAuthService.retrieveSgidUserEmails(accessToken, sub)
       )
-      .andThen((email) => {
-        if (!email || !email.endsWith("@open.gov.sg")) {
+      .andThen((officerDetails) => {
+        if (!officerDetails || officerDetails.length === 0) {
           return errAsync(
             new AuthError(
               "Your email has not been whitelisted to use sgID login. Please use another method of login."
             )
           )
         }
-        return okAsync(email)
+        return okAsync(officerDetails)
       })
-      .andThen((email) =>
-        // Login with email
+    if (userDataRes.isErr()) {
+      const { error } = userDataRes
+      if (
+        error instanceof SgidFetchAccessTokenError ||
+        error instanceof SgidFetchUserInfoError
+      ) {
+        return res.status(500).send()
+      }
+      if (error instanceof AuthError) {
+        return res.status(401).send()
+      }
+      return res.status(500).send()
+    }
+
+    const userData = userDataRes.value
+    if (userData.length === 1) {
+      // If user only has a single email, login directly
+      await ResultAsync.fromPromise(
+        this.usersService.loginWithEmail(userData[0].email),
+        (error) => {
+          logger.error(
+            `Error while retrieving user info from database: ${error}`
+          )
+          return new DatabaseError()
+        }
+      )
+        .map((user) => {
+          const userInfo = {
+            isomerUserId: user.id,
+            email: user.email,
+          }
+          Object.assign(req.session, { userInfo })
+          return res.status(200).send({ userData })
+        })
+        .mapErr((error) => res.status(500).send())
+    } else {
+      // User has multiple emails, defer to user to select email
+      const csrfTokenExpiry = new Date()
```

Comment:
Why do we not require the same expiring and signing for single email users?

Reply:
The expiring/signing is to ensure that a user attempting to use the multiuser login actually has access to that user - at this point we actually are able to log in the user as any of their approved emails already. But since we don't know which user they want to use, this requires additional input from the user, which requires us to ensure that an attacker cannot simply call this endpoint pretending to be the user.

Reply:
don't they hit this point **only after** they pass authorization checks from sgid? 

the multi-user email selection flow seems to be covered below

Reply:
oh ok just double checking - this authorizes and retrieves data from sgid then passes back to FE

user selects email on FE and hits the below endpoint, **which verifies the security token** to ensure identity

if this is the flow, then please resolve